### PR TITLE
2.7: Changes to Auto-configuration #156

### PR DESCRIPTION
**Pull request by ijusti**

### CODE REVIEW

File: components/sbm-core/src/main/java/org/springframework/sbm/build/impl/OpenRewriteMavenBuildFile.java [hunk 1]
```diff
@@ -345,7 +366,7 @@ protected void addDependenciesInner(List<Dependency> dependencies) {
             apply(r);
             apply(new RefreshPomModel());
             List<Dependency> exclusions = dependencies.stream()
-                    .filter(d -> false == d.getExclusions().isEmpty())
+                    .filter(not(d -> d.getExclusions().isEmpty()))
```
Comment (fabapp2) [inline components/sbm-core/src/main/java/org/springframework/sbm/build/impl/OpenRewriteMavenBuildFile.java:369]: TIL

**Comment by fabapp2** [comment]
Hi @ijusti! 
I've been on PTO thus the long delay.
Thanks a lot for your contribution and the many beautifications and fixes. Very much appreciated! 🙏 
I had a few comments regarding the operation on files and would ask you to provide at least a test for the `Action` itself.
If you could resolve these comments I'll happily merge your PR.
Thanks again!

File: components/sbm-recipes-boot-upgrade/src/main/resources/recipes/boot-autoconfiguration-update.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+- name: boot-autoconfiguration-update
+  description: Create org.springframework.boot.autoconfigure.AutoConfiguration.imports file for new spring 2.7
+  condition:
+    type: org.springframework.sbm.common.migration.conditions.TrueCondition
```
Comment (fabapp2) [inline components/sbm-recipes-boot-upgrade/src/main/resources/recipes/boot-autoconfiguration-update.yaml:4]: The condition for the recipe should better be `IsMatchingSpringBootVersion` with pattern `2.7.` so it is only shown for 2.7 Boot applications.

**Review by fabapp2 — CHANGES_REQUESTED** [review]

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import static org.springframework.sbm.boot.upgrade.common.actions.CreateAutoconfigurationAction.getAutoConfigurationPath;
+
+public class BootHasAutoconfigurationCondition implements Condition {
+    @Override
+    public String getDescription() {
+        return "Check if there is autoconfiguration";
+    }
+
+    @Override
+    public boolean evaluate(ProjectContext context) {
```
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java:37]: i need to check that near spring.factories no new file exists

**Review by fabapp2 — CHANGES_REQUESTED** [review]

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public static Path getAutoConfigurationPath(Path springFactoriesPath) {
+        return Path.of(springFactoriesPath.getParent() + AUTO_CONFIGURATION_IMPORTS);
+    }
+
+    public static boolean isSpringFactory(Path path) {
+        return path.endsWith(SPRING_FACTORIES_PATH);
+    }
+
+    @Override
+    public void apply(ProjectContext context) {
```
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java:49]: this search does not work. there is no spring.factories in ProjectContext. That is why I worked directly with files

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java:37]: this good, and I tried to make this way? but there is no spring.factories in context

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java:49]: Also I guess parser should also understand new file 
`**/src/main/resources/META-INF/org.springframework.boot.autoconfigure.AutoConfiguration.imports`

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java:37]: something like this, that mean that there is no new imports near spring.factories

```
return context.search(new PathPatternMatchingProjectResourceFinder("**/src/main/resources/META-INF/spring.factories"))
.stream()
.anyMatch(resource-> context.search(new PathPatternMatchingProjectResourceFinder("resource.path.parent +"org.springframework.boot.autoconfigure.AutoConfiguration.imports")).isEmpty())
```

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabapp2) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java:49]: The parser is fixed now (PR #191) and yes, the new resource should be parsed by the parser.

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabapp2) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/como/conditions/BootHasAutoconfigurationCondition.java:37]: Ah, I got your valid point here.
I am thinking that if we replace all old with new factories in one go and remove the old ones (or move with `MoveFilesAction`), there should not be both (old and new) in one project. If this is "guaranteed" we can skip the second part of the check?

You could replace `resource.path.parent` with `resource.getAbsoluetPath().getParent().resolve("org.springframework.boot.autoconfigure.AutoConfiguration.imports")` and then use `AbsolutePathResourceFinder` with this path.

**Comment by fabapp2** [comment]
Hi @ijusti !
Could you provide the requested changes?
I am sorry if I am pushy here but we'll need this issue (very) soon.
It's no problem if you don't have time, I'd merge what you did and add the requested changes myself then (crediting you as author). 
Let me know what works for you?

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ijusti) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java:49]: after rebase scan command always has **projectRoot="."**
image

[the remark this answers is not in the document]
Comment (fabapp2) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade/common/actions/CreateAutoconfigurationAction.java:49]: Hi @ijusti !
You're right with the recent update to spring-shell `2.1.0-RC1` the command property resolution changed.
You'd either need to add `arity=1` or provide the name of the property like so `scan --projectRoot=...`.
As we need this change today, my colleagues will finish it. I will merge back what you provided so far and they'll add the test, condition, and refactored action. We will mention you as a contributor though.

**Pull request merged**